### ISSUE TITLE
Add web socket interface and browser block explorer.

### ISSUE THREAD

**Issue by evoskuil** [libbitcoin/libbitcoin-server#503]

**Comment by thecodefactory** [comment]
I've got a document in the works going over a couple simple command line tools for websocket query and the web interface I've been working with (which was expected to be a separate tree clone).  I don't think the interface should be official in any sense, so how about I move the web (root dir) and query examples into libbitcoin-server/examples and move the document in the works to the server wiki?

**Comment by evoskuil** [comment]
> I don't think the interface should be official in any sense

Not sure I understand, are you referring to the websockets interface?

**Comment by thecodefactory** [comment]
> The idea is for the server to serve this website whenever its running with the endpoint enabled.

Oh!  I didn't realize that was the intent of the test interface I made.  I used it for testing websocket functionality while building that.  I expected people would build their own (dare I say, professional) UI interfaces, using our exposed websocket functionality.  I'm ok including it, but I can't say it's much more than an example.  Also, any web UI update would then be commits against server, where it may be better to have something like libbitcoin-web where the style or format changes are made independently and aren't tied to server releases.  As it is, the web root directory is configurable.

**Comment by evoskuil** [comment]
Doesn't have to be pretty, somebody will come along and take care of that later :). But should provide a usable GUI into the server/node. Eventually we want to make this into a web admin interface.

**Comment by thecodefactory** [comment]
> >I don't think the interface should be official in any sense

> Not sure I understand, are you referring to the websockets interface?

2 things here, sorry for the wording confusion.  When I think of the websocket interface, I am thinking of the websocket API exposed by libbitcoin server web services.

That's obviously confusing if you think of the websocket interface being the web UI visualized in a browser.  The former is merged.  The second is in doc stages and a private repo.


> Doesn't have to be pretty, somebody will come along and take care of that later :). But should provide a usable GUI into the server/node. Eventually we want to make this into a web admin interface.

Ok.  I strongly advocate for a separate repo though, like libbitcoin-web (or whatever).  Any thoughts on that?   Even an admin interface would be using some web (HTTP/JSON-RPC/websocket) interface, so it's safe to divide by repo line.

So it could naively be something like:
```
# cd to directory where server or node runs
ln -s ../libbitcoin-web web
# startup like normal
```

I know that's an extra step for people to miss, but separating web UI work from server is important, as I expect it'll be more active in the future.  Perhaps @pmienk knows a way to integrate detection of the repo and automatic linking/installation?

**Comment by evoskuil** [comment]
> if you think of the websocket interface being the web UI visualized in a browser

that's just a web page. I'd refer to it as a GUI or website, not an API or "interface".

**Comment by thecodefactory** [comment]
> I'd like to have a simple default site integrated into the server

How about a page that describes how to setup a web root? ;-)

> When we build the server these pages should be part of the single-file statically-linked binary. IMO that aspect will just be complicated by another repo.

Ok, but I think this is a mistake.  Every minor web change requires a new binary build/release, which is not how web developers are used to working, given modern environments.  I do think it lends itself well to encouraging people to not use (or build on) the default one though.  Maybe back to my first question above?

**Comment by evoskuil** [comment]
Your thinking of the build as producing a developer environment, which is understandable because you are a developer.

**Comment by evoskuil** [comment]
Also you are making an assumption about the build that isn’t what I’m implying. A repo is not available to people until they go get it. This requires a path config, which we can/do support. But the node should work by default as well, with no required config or external components.

**Comment by thecodefactory** [comment]
> You are thinking of the build as producing a developer environment, which is understandable because you are a developer.

Guilty :-)

But the point of exposing the web/JSON-RPC/websocket API is for developers to use them.  I have no doubt there are some great web devs out there that may find and take ownership of the UI aspects, which would be ideal.

I also have past experience with tying web code with binary release code and made production runs hell until they were split up (regarding versioning and deployment, etc).  I agree that if it's a UI that never changes, there's no issue.  But it usually takes a while to get there.

**Comment by evoskuil** [comment]
The GUI is not intended as a developer tool, it’s an admin console to the application. Like the command line interface it’s meant to be replaced when a developer builds a custom server on libbitcoin. The websocket interface is exposed via a port, and is not affected.

The build is automated, so it only requires a pull request to change the presentation, which would be the case in another repo. And any GUI development should take place in a configured directory, which can of course be pointed to same or another repo.

**Comment by thecodefactory** [comment]
Ok, to be sure we're on the same page, how about this:  Let me finish my document and then you give it a quick run through to get the web interface up.  When you're done, if you still want me to plop the contents of the separate repo into libbitcoin-server/data/web (or wherever), I'll do it then and we can call it a day on this?

**Comment by thecodefactory** [comment]
This is still in rough shape, but web stuff is documented here for now: https://github.com/thecodefactory/libbitcoin-web

**Comment by evoskuil** [comment]
In the server I'm not looking for a tutorial, examples or full exercise of the web/sockets interface. I'm just looking to expose the server's blockchain via a simple embedded block explorer. Tutorials and example code should go in the server wiki, just as with the zeromq interface and docs for all of the libs.

**Comment by thecodefactory** [comment]
This should have been closed when #508 was merged.
